src/main/kotlin/org/elm/lang/core/types/TypeReplacement.kt

```diff
@@ -159,21 +176,33 @@ class TypeReplacement(
             fieldReferences: RecordFieldReferenceTable,
             wasMutable: Boolean
     ): Ty {
-        val replacedBase = if (baseTy == null || baseTy !is TyVar) null else getReplacement(baseTy)
-        val newBaseTy = when (replacedBase) {
-            // If the base ty of the argument is a record, use it's base ty, which might be null.
-            is TyRecord -> replacedBase.baseTy
+        val oldBase = if (baseTy == null || baseTy !is TyVar) null else getReplacement(baseTy)
+        val newBase = when (oldBase) {
+            // If the base ty of the argument is a record, use its base ty, which might be null.
+            is TyRecord -> oldBase.baseTy
             // If it wasn't substituted, leave it as-is
-            null -> baseTy
+            null -> when {
+                // Although if it's a record, we might need to make it immutable
+                baseTy is MutableTyRecord && !keepRecordsMutable -> replace(baseTy)
+                else -> baseTy
+            }
             // If it's another variable, use it
-            else -> replacedBase
+            else -> oldBase
         }
 
-        val declaredFields = fields.mapValues { (_, it) -> replace(it) }
-        val baseFields = (replacedBase as? TyRecord)?.fields.orEmpty()
-        val baseFieldRefs = (replacedBase as? TyRecord)?.fieldReferences
+        val baseFields = (oldBase as? TyRecord)?.fields.orEmpty()
+        val baseFieldRefs = (oldBase as? TyRecord)?.fieldReferences
 
-        var newFieldReferences = when {
+        // Make the new map as small as we can for these fields. HashMap always uses a power of two
+        // element array, so we can't avoid all wasted memory.
+        val initialCapacity = baseFields.size + fields.keys.count { it !in baseFields }
+        val newFields = LinkedHashMap<String, Ty>(initialCapacity, 1f)
+
+        // Don't use putAll here, since that function will resize the table to hold (size + 1) elements
```

Comment:
You had a similar comment for `RecordFieldReferenceTable`. Maybe it's not possible due to different types, but would it be possible to define an extension method that makes the code self-documenting and reusable in both places?

Reply:
They're different types, so we'd need to functions. But it turns out that we can use `addAll` on a `SmartHashSet` after all, so I'll remove the comment there.